pinot-spi/src/main/java/org/apache/pinot/spi/utils/CommonConstants.java

```diff
@@ -248,6 +248,9 @@ public static class Instance {
     public static final String CONFIG_OF_MULTI_STAGE_ENGINE_MAX_SERVER_QUERY_THREADS =
         "pinot.beta.multistage.engine.max.server.query.threads";
     public static final String DEFAULT_MULTI_STAGE_ENGINE_MAX_SERVER_QUERY_THREADS = "-1";
+    public static final String CONFIG_OF_MULTI_STAGE_ENGINE_MAX_SERVER_QUERY_HARDLIMIT_FACTOR =
+        "pinot.beta.multistage.engine.max.server.query.threads.hardlimit.factor";
+    public static final String DEFAULT_MULTI_STAGE_ENGINE_MAX_SERVER_QUERY_HARDLIMIT_FACTOR = "4";
```

Comment:
I feel a fixed number would be easier to understand.
With current behavior, if user somehow put `-1` as the factor because they want to disable this, it will end up enabling the hard limit with 1 thread

Reply:
I see your point, but kind of resist to not having a way to make that factor configurable. I'd rather see how this performs in real scenarios and have that option in case we need to fine-tune it.

So far, I added an (ugly) check to avoid setting a hard limit in case thread limit is already non-positive.